### PR TITLE
AArch64: Implement Interface Call Dynamic PIC

### PR DESCRIPTION
**Pull request by Akira1Saitoh**
This commit implements dynamic pic for interface call.
The design is based on ppc's implementation.

Depends on https://github.com/eclipse/omr/pull/6126

Co-authored-by: KONNO Kazuhiro <konno@jp.ibm.com>
Signed-off-by: Akira Saitoh <saiaki@jp.ibm.com>

### CODE REVIEW

**Comment by knn-k** [comment]
jenkins test sanity alinux64 jdk11 depends eclipse/omr#6126

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp [hunk 1]
```diff
@@ -1520,16 +1520,51 @@ static void buildVirtualCall(TR::CodeGenerator *cg, TR::Node *callNode, TR::Regi
  *
  * @param[in] cg:          code generator
  * @param[in] callNode:    node for the interface call
+ * @param[in] vftReg:      vft register
+ * @param[in] tmpReg:      temporary register
```
Comment (0xdaryl) [inline runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp:1524]: Spelling on line below:  interface

File: runtime/compiler/aarch64/runtime/PicBuilder.spp [hunk 2]
```diff
@@ -738,15 +743,41 @@ const_movz_x9:
 //
 // in:     x30 = snippet
 //
-// trash:	x10, x11
+// trash:	x8, x10, x11
+//
+// At the initial state, the `bl` instruction in the Interfae Call Snippet points to `_interfaceCallHelper`,
```
Comment (0xdaryl) [inline runtime/compiler/aarch64/runtime/PicBuilder.spp:748]: sp:  Interface

**Review by 0xdaryl — COMMENTED** [review]
Just a couple of minor spelling mistakes.  We don't need to re-run CI just for that.

I do want to review the assembly code one more time before merging.

File: runtime/compiler/runtime/Runtime.cpp [hunk 3]
```diff
@@ -1579,7 +1580,8 @@ void initializeCodeRuntimeHelperTable(J9JITConfig *jitConfig, char isSMP)
    SET(TR_ARM64interpreterDoubleStaticGlue,       (void *) _interpreterDoubleStaticGlue,     TR_Helper);
    SET(TR_ARM64interpreterSyncDoubleStaticGlue,   (void *) _interpreterSyncDoubleStaticGlue, TR_Helper);
    SET(TR_ARM64nativeStaticHelper,                (void *) _nativeStaticHelper,              TR_Helper);
-   SET(TR_ARM64interfaceDispatch,                 (void *) _interfaceDispatch,               TR_Helper);
+   SET(TR_ARM64interfaceCompleteSlot2,             (void *) _interfaceCompleteSlot2,           TR_Helper);
```
[the remark this answers is not in the document]
Comment (Akira1Saitoh) [reply runtime/compiler/runtime/Runtime.cpp]: I fixed it.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp (hunk 1 above)
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp:1524]: Fixed it.

File: runtime/compiler/aarch64/runtime/PicBuilder.spp (hunk 2 above)
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/runtime/PicBuilder.spp]: Thanks, I fixed it.

**Comment by Akira1Saitoh** [comment]
sanity.functional, sanity.system and sanity.openjdk passed on internal Jenkins.
`job/Pipeline-Build-Test-Personal/10024/`
`job/Grinder/17238/`
`job/Grinder/17246/`

**Comment by knn-k** [comment]
jenkins compile alinux64 jdk11 depends eclipse/omr#6126

**Review by 0xdaryl — COMMENTED** [review]
I completed a second review.  Found a couple of typos only.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp [hunk 4]
```diff
@@ -1821,9 +1856,15 @@ void J9::ARM64::PrivateLinkage::buildVirtualDispatch(TR::Node *callNode,
             if (methodSymbol->isInterface())
                {
                TR::LabelSymbol               *ifcSnippetLabel = generateLabelSymbol(cg());
-               TR::ARM64InterfaceCallSnippet *ifcSnippet = new (trHeapMemory()) TR::ARM64InterfaceCallSnippet(cg(), callNode, ifcSnippetLabel, argSize, doneOOLLabel, static_cast<uint8_t *>(thunk));
+               TR::LabelSymbol               *firstClassCacheSlotLabel = generateLabelSymbol(cg());
+               TR::LabelSymbol               *firstBranchAddressCacheSlotLabel = generateLabelSymbol(cg());
+               TR::LabelSymbol               *secondClassCacheSlotLabel = generateLabelSymbol(cg());
+               TR::LabelSymbol               *secondBramchAddressCacheSlotLabel = generateLabelSymbol(cg());
```
[the remark this answers is not in the document]
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: Thanks. I fixed it.

File: runtime/compiler/aarch64/runtime/PicBuilder.spp [hunk 5]
```diff
@@ -738,15 +743,41 @@ const_movz_x9:
...
 	stp	x5, x4, [J9SP, #16]
 	stp	x3, x2, [J9SP, #32]
 	stp	x1, x0, [J9SP, #48]
 	mov	x7, x30						// preserve LR
-	ldr	x10, [x30, #J9TR_ICSnippet_codeCacheReturnAddress]	// protect code cache RA in x10 (in L_commonLookupException, it is expected)
+	ldr	x10, [x30, #J9TR_ICSnippet_codeCacheReturnAddress]	// protect code cache RA in x10 (in L_commonJitDispatch, it is expected)
+
+	ldr	x0, [x30, #J9TR_ICSnippet_FirstClass]		// first cache class
+	cmp	x0, #-1
+	bne	L_continueLookup				// If first cache class it not -1, resolve is already done
```
[the remark this answers is not in the document]
Comment (Akira1Saitoh) [reply runtime/compiler/aarch64/runtime/PicBuilder.spp]: I fixed it.

**Comment by 0xdaryl** [comment]
Previous CI passed plus internal testing.  I don't think there is a need to re-run full CI on this latest change (typos only), so I will just launch an AArch64 build.

**Comment by 0xdaryl** [comment]
Jenkins test sanity alinux64 jdk11

**Comment by 0xdaryl** [comment]
Jenkins test sanity alinux64 jdk11

**Comment by 0xdaryl** [comment]
Those same 4 failures occurred twice in successive test runs.  I don't believe they are related to this PR, but might be an infra issue on the new AArch64 PR build machine because we did not see them before with earlier testing on this same PR.

**Comment by knn-k** [comment]
@Akira1Saitoh Can you rebase your commit for this PR, please?  Thanks.

**Comment by Akira1Saitoh** [comment]
@knn-k Okay, I rebased the commit.

**Comment by knn-k** [comment]
Jenkins test sanity alinux64 jdk11

**Comment by knn-k** [comment]
Four tests failed again after rebasing:
- cmdLineTester_GCCheck_0
- cmdLineTester_GCCheck_1
- cmdLineTester_callsitedbgddrext_openj9_0
- cmdLineTester_pltest_0

**Comment by 0xdaryl** [comment]
This PR has been tested both at the OpenJ9 project and internally.  The four CI failures appear due to to the inability to produce core files on new AArch64 PR build machine and have run successfully in the past.  Merging.

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**